services/outputhost/extcache.go

```diff
@@ -237,15 +236,28 @@ func (extCache *extentCache) loadReplicaStream(startAddress int64, startSequence
 
 		// First try to start from the already set offset in metadata
 		if startAddress == 0 {
-			// If consumer group wants to start from a timestamp, get the address from the store
-			startFrom := extCache.startFrom
+			// If consumer group wants to start from a timestamp, get the address from the store.
+			// Note: we take into account any 'delay' that is associated with the CG and the 'skip-older'
+			// time that was specified, by offsetting the time appropriately. we apply the 'start-from'
+			// and 'skip-older' to the "visibility time" (enqueue-time + delay) of the message as opposed
+			// to the enqueue-time.
 
 			// TODO: if the consumer group wants to start from the beginning, we should still calculate the earliest address
 			//       that they can get. To do otherwise means that will will get spurious 'skipped messages' warnings
 			// NOTE: audit will have to handle an uneven 'virtual startFrom' across all of the extents that a zero startFrom
 			//       consumer group is reading
 			// NOTE: there is a race between consumption and retention here!
-			if startFrom > 0 {
+			if extCache.startFrom.UnixNano() > 0 || extCache.skipOlder > 0 {
+
+				var startFrom int64
+
+				// compute start-from as the max(skip-older-time, start-from), adjusting for 'delay', if any
+				if time.Now().Add(-extCache.skipOlder).After(extCache.startFrom) {
+					startFrom = time.Now().Add(-extCache.skipOlder).Add(-extCache.delay).UnixNano()
```

Comment:
we adjust for the 'delay' proportionally .. on both 'startFrom' and 'now - skip-older'; so the if-check only needs to find out which of them is greater .. and then apply the delay to whichever was the chosen one. i thought the comment did imply that ..